Update Staking Validators Rest Query

alexanderbez opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

Support pagination and bond status query params in the `/staking/validators` REST endpoint. The default number of validators  returned is the max number of bonded validators and the default status is bonded.

closes: #4099 

----

- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/develop/CONTRIBUTING.md#pr-targeting))

- [x] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [x] Updated relevant documentation (`docs/`)
- [x] Added a relevant changelog entry: `sdkch add [section] [stanza] [message]`
- [x] rereviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

fedekunze commented on client/lcd/swagger-ui/swagger.yaml:
```diff
@@ -764,6 +764,22 @@ paths:
   /staking/validators:
     get:
       summary: Get all validator candidates
```
```suggestion
      summary: Get all validator candidates. By default it returns only the bonded validators.
```

fedekunze commented on client/lcd/swagger-ui/swagger.yaml:
```diff
@@ -764,6 +764,22 @@ paths:
+        - in: query
+          name: status
+          type: string
+          description: The validator bond status. Must be either bonded, unbonded, or unbonding.
```
```suggestion
          description: The validator bond status. Must be either 'bonded', 'unbonded', or 'unbonding'.
```

fedekunze commented on x/staking/querier/querier_test.go:
```diff
@@ -134,19 +132,28 @@ func TestQueryValidators(t *testing.T) {
 	queriedValidators := keeper.GetValidators(ctx, params.MaxValidators)
 
-	res, err := queryValidators(ctx, cdc, keeper)
+	queryValsParams := NewQueryValidatorsParams(1, int(params.MaxValidators), sdk.BondStatusUnbonded)
```
can we add another test for the `BondStatusUnbonding` ?

alexanderbez replied:
Sure, I can see if I can manually set their status.

alexanderbez replied:
Updated.

rigelrozanski commented on types/staking.go:
```diff
@@ -25,17 +25,21 @@ const (
 
-//BondStatusToString for pretty prints of Bond Status
+// BondStatusToString for pretty prints of Bond Status.
 func BondStatusToString(b BondStatus) string {
```
We should modify this to be `func (b BondStatus) String() string`

alexanderbez replied:
Agreed. I'll update.

rigelrozanski commented on x/staking/querier/querier.go:
```diff
@@ -128,14 +129,47 @@ func NewQueryRedelegationParams(delegatorAddr sdk.AccAddress, srcValidatorAddr s
 	stakingParams := k.GetParams(ctx)
-	validators := k.GetValidators(ctx, stakingParams.MaxValidators)
+	if params.Limit == 0 {
+		params.Limit = int(stakingParams.MaxValidators)
```
So `AllValidators` can be greater than the `MaxValidators` (as it contains bonded and non-bonded validators) - maybe we want to set the page limit to 100 by default but it seems kinda arbitrary. It would be my preference to simply error at a page limit of 0. Can we set default page limits elsewhere in rest?

alexanderbez replied:
So the REST utils actually does have a default limit -- 30. This is why I manually override it to 0. I wanted to keep this endpoint compatible with it's most intended use -- give me all bonded validators. So clients may omit sending `page` and `limit` and get the expected results.

I don't think we should error when no page or limit is provided.

jackzampolin replied:
agree here

rigelrozanski replied:
👍

rigelrozanski commented on x/staking/querier/querier.go:
```diff
@@ -128,14 +129,47 @@ func NewQueryRedelegationParams(delegatorAddr sdk.AccAddress, srcValidatorAddr s
-	res, errRes := codec.MarshalJSONIndent(cdc, validators)
+	if start >= len(filteredVals) {
+		// page is out of bounds
+		filteredVals = []types.Validator{}
```
should we not directly error here? (- if we did then we unindent everything in the following else statement)

alexanderbez replied:
Hmmm, I don't think. Asking for a page beyond the total should just give you no results. This is generally how I've seen pagination work (eg. Tendermint also does this).

rigelrozanski replied:
👍

alexanderbez commented:
Thanks for the review @rigelrozanski -- I addressed all your comments. With regards to the CLI, that was left untouched because the CLI and REST never implemented the same functionality really (ie. the CLI just returns _all_ validators _always_). Should we change this too?
